### DISCUSSION TITLE
Reading png images and labels in Monai

### DISCUSSION THREAD

**Discussion by Harris91-lee** [Project-MONAI/MONAI#7479] [General]
Hello everyone, looks like LoadImaged is used for .nii.gz 3D volume: LoadImaged(keys = ["image", "label"]), how can i do that for only 2D png images?

**Comment by KumoLiu** [comment]
Hi @Harris91-lee, we have `PILReader` which can be used in `LoadImage`.
https://github.com/Project-MONAI/MONAI/blob/ff430286c37e78d7592372a5a97377f0cbb0219c/monai/data/image_reader.py#L1129

Hope it helps.

**Reply by Harris91-lee** [reply]
Thank you a lot @KumoLiu ^^ 
can you please help me reading image and label using the dataloader after applying augmentation? Currently im trying to figure this out. Previously in the 3D: **augmentation = Compose([LoadImaged(keys = ["image", "label"]),**
however now as you suggested im going to replace LoadImaged with **LoadImaged(keys=["image", "label"], image_only=False, reader=PILReader(converter=lambda image: image.convert("L"))),**      
Am I doing okay?

**Reply by KumoLiu** [reply]
`reader="PILReader"` should be ok.
https://github.com/Project-MONAI/MONAI/blob/ff430286c37e78d7592372a5a97377f0cbb0219c/monai/transforms/io/array.py#L138-L144

**Reply by Harris91-lee** [reply]
Thank you SO much for your fast reply @KumoLiu  , you are a life saver. 
so the correct form is: LoadImaged(keys=["image", "label"], image_only=False, reader="PILReader", reverse_indexing=False), and if i load only images no label: LoadImaged(keys="image", image_only=True, reader="PILReader", reverse_indexing=False)?

**Reply by KumoLiu** [reply]
Yes, feel free to ask if meet any issue. Thanks.

**Reply by Harris91-lee** [reply]
so there is something wrong in loading the images with their labels, the images tesnor shape is  **[1,1,512,512]** which is as expected but the label comes like **[1,512,512,512]** though im using the same process  for both: 
class Custom_Dataset(CacheDataset):
    def __init__(self, data_path, is_label = True, select_label = None, crop_size = None , input_size = None,  augmentationMode = None, cache_rate = 0):
        dt_data = open(data_path, 'r').readlines()
        dt_data = [l.strip().split() for l in dt_data]
        self.filenames_input = [os.path.join(data_dir, item[0]) for item in dt_data]
        self.is_label = is_label
        if is_label:
            self.filenames_label = [os.path.join(data_dir, item[1]) for item in dt_data]  
            self.select_label = select_label   # here it says true. 

        self.crop_size = crop_size
        self.input_size = input_size

        data_files = self.__datafiles__()
        augmentation = self.__augmentation__(augmentationMode)
        super().__init__(data = data_files, transform = augmentation, cache_rate = cache_rate)

    # 1. setup database
    def __datafiles__(self):
        filenames_lst = []
        for file_name_dir in self.filenames_input:
            file_name = os.path.basename(file_name_dir)
            # file_name = file_name.replace('.nii.gz', '')
            file_name = file_name.replace('_image.png', '')

            filenames_lst.append(file_name)

        if self.is_label:
            data_files = [{"filename": filename, "image": img, "label": seg} for filename, img, seg in zip(filenames_lst, self.filenames_input, self.filenames_label)]
        else:
            data_files = [{"filename": filename, "image": img} for filename, img in zip(filenames_lst, self.filenames_input)]
        
        return data_files
        
        Do you see where is my mistake @KumoLiu ?

**Reply by Harris91-lee** [reply]
Oh so i solved it by doing EnsureChannelFirstd(keys=["image", "label"]), instead EnsureChannelFirstd(keys="image"),

**Comment by Harris91-lee** [comment]
Hello Mr YunLiu  @KumoLiu, hope you are doing well :) 
i have a further question regarding the  **LoadImaged(keys=["image", "label"], image_only=False, reader="PILReader")** how can i enforce reading images in gray so channel will be always 1? currently my data is mixed of gray and RGB so i have conflict regarding that. 
Thank you so much in advance.

**Reply by Harris91-lee** [reply]
converting the images to gray externally is a solution i know but i wonder if i can do it internally ^_^

**Reply by KumoLiu** [reply]
Hi @Harris91-lee,
There isn't a direct transform function for converting RGB images to grayscale.
However, you can create a custom transform using common Python libraries such as OpenCV or PIL, and integrate it into your MONAI pipeline.
Here is a simple example of a custom MONAI transform:
```
from monai.transforms import Transform
from PIL import Image

class RgbToGray(Transform):

    def __call__(self, img):
        gray_img = Image.fromarray(img.astype('uint8')).convert('L')
        return np.array(gray_img)
```
The Transform base class is leveraged here to create a custom transform that can be integrated into a MONAI workflow. In this case, the transformation converts an RGB image to a grayscale image using the PIL library's Image.convert('L') method. This custom transform can then be incorporated into your MONAI pipeline like any other MONAI transform.
Please keep in mind that this transform assumes that data coming in are numpy arrays of image data. Depending on what your actual pipeline looks like, you might need to adjust the transform.

Hope it helps, thanks.

**Reply by Harris91-lee** [reply]
Thank you  @KumoLiu. 
that was indeed a great way, will use it ^^ 
During the training process i noticed my model size is so huge close to 1GB, is it normal to have this big size model? Are there any options I should add/remove if i want to reduce the model size? 
I'm so grateful for your help and support though.

**Comment by Harris91-lee** [comment]
Hello @KumoLiu, hope you are doing well? 
go tanother question regarding the **LoadImaged(keys=["image"], image_only=True, reader="PILReader")** 
i noticed it does random/90 degree rotation to the data, is that normal? and how to stop that from happening? 
many thankls in advance. 
- Harris

**Reply by KumoLiu** [reply]
Hi @Harris91-lee, you can set `reverse_indexing=False`, this is enabled by default, so that output of the reader is consistent with the other readers.
https://github.com/Project-MONAI/MONAI/blob/e9e273858b6904659b0cd581d357662dc5c782f6/monai/data/image_reader.py#L1136-L1138

Thanks.

**Reply by Harris91-lee** [reply]
Thank you @KumoLiu  :)
you are right, after adding the reverse_indexing=False no rotation happens but surprisingly the model result got worsen as the train dataloader didnt have reverse_indexing=False. Thought it's okay if the rotation happened in the training but seems i should unify the LoadImaged for both right?

**Reply by Harris91-lee** [reply]
So when i unify the train and val LoadImaged to have **reverse_indexing=False** the result comes out fine.
Thank you again @KumoLiu ~
